src/dmd/expression.d

```diff
@@ -5307,6 +5307,49 @@ extern (C++) final class DotVarExp : UnaExp
         if (e1.op == TOK.this_)
             return var.checkModify(loc, sc, e1, flag);
 
+        // Issue 12764 : https://issues.dlang.org/show_bug.cgi?id=12764
+
+        // If inside a constructor and an expression of type this.field.var
+        // is encountered, where field is a struct declaration with
+        // default construction disabled, we must check if that field was
+        // properly initialized before being able to actually use its
+        // fields.
+        if (sc.func)
+        {
+            auto ctd = sc.func.isCtorDeclaration();
+
+            // if inside a constructor scope and e1 of this DotVarExp
+            // is a DotVarExp, then check if e1.e1 is a `this` identifier
+            if (ctd && e1.op == TOK.dotVariable)
+            {
+                scope dve = cast(DotVarExp)e1;
+                if (dve.e1.op == TOK.this_)
+                {
+                    scope v = dve.var.isVarDeclaration();
+                    // if v is a struct member field with no initializer and no default construction
+                    if (v && v.isField() && v.type.ty == Tstruct && !v._init)
+                    {
+                        const sd = (cast(TypeStruct)v.type).sym;
+                        if (sd.noDefaultCtor)
+                        {
+                            // check if v was initialized. if not we must error
+                            const ad = ctd.isMember2();
+                            size_t i;
+                            for (i = 0; i < sc.ctorflow.fieldinit.length; i++)
```

Comment:
I think we need an `Array.indexOf` method.

Reply:
-> https://github.com/dlang/dmd/pull/8228